forge/ee/routes/sso/auth.js

```diff
@@ -87,12 +96,66 @@ module.exports = fp(async function (app, opts) {
                 }
                 done(null, user)
             } else {
-                const unknownError = new Error(`Unknown user: ${samlUser.nameID}`)
-                unknownError.code = 'unknown_sso_user'
-                const userInfo = app.auditLog.formatters.userObject({ email: samlUser.nameID })
-                const resp = { code: 'unknown_sso_user', error: 'unauthorized' }
-                await app.auditLog.User.account.login(userInfo, resp, userInfo)
-                done(unknownError)
+                const state = JSON.parse(request.body.RelayState)
+                const providerOpts = await app.sso.getProviderOptions(state.provider)
+                if (providerOpts.provisionNewUsers) {
+                    // create new user from content of samlUser if available
+                    const userProperties = {
+                    //     username: request.body.username,
+                    //     name: request.body.name,
+                        email: samlUser.nameID,
+                        // email_verified: true,
+                        // password: request.body.password,
+                        admin: false,
+                        // sso_enabled: true,
+                        tcs_accepted: new Date()
+                    }
+
+                    if (samlUser['http://schemas.microsoft.com/identity/claims/displayname']) {
+                        userProperties.name = samlUser['http://schemas.microsoft.com/identity/claims/displayname']
+                    } else {
+                        userProperties.name = samlUser.nameID.split('@')[0]
+                    }
+
+                    userProperties.password = generatePassword()
+                    userProperties.username = samlUser.nameID.replace('@', '-').replaceAll('.', '_')
+
+                    try {
+                        // create user
+                        const newUser = await app.db.models.User.create(userProperties)
+
+                        // check if we need to add teams from SSO
+                        if (providerOpts.groupMapping) {
+                            // This SSO provider is configured to manage team membership.
+                            try {
+                                await app.sso.updateTeamMembership(samlUser, newUser, providerOpts)
```

Comment:
If the SSO provider is configured to manage team membership, we should *not* create the trial team for them - as they will be removed from that team instantly (or at least, might be...)

Reply:
ahh, OK, that will make a bunch of things cleaner.

Means I can also remove the new file/shared function

Reply:
@knolleary but this does mean it would need to move the provision users switch into the enable group management section.

And we don't have group membership for LDAP.